### ISSUE TITLE
Error When Export YOLO-Pose Annotations

### ISSUE THREAD

**Issue by kenwaytis** [CVHub520/X-AnyLabeling#471]
### System Information:
1. Ubuntu 22.04
2. X-AnyLabeling 2.3.6 CPU
3. Started using the source code, exporting YOLO object detection format works fine

---

### Issue Description:
When I completed the image annotation with rotated boxes and keypoints, I followed the user_guide to write `pose.yaml`. After selecting and choosing `pose.yaml`, **an error occurred**: `int() argument must be a string, a bytes-like object or a number, not 'NoneType'`.

### Additional Information:
1. The tags I annotated are as shown in the image, with "instrument" being the detection box and the rest being keypoints.
![2024-06-24_10-29](https://github.com/CVHub520/X-AnyLabeling/assets/40118038/f1be1fc1-5a9f-4e89-abda-670472147e6a)
2. The `pose.yaml` I wrote is as follows:
```yaml
has_visible: false
classes:
    instrument:
      - center
      - point
      - min
      - max
```
3. Screenshot of error:
![2024-06-24_10-43](https://github.com/CVHub520/X-AnyLabeling/assets/40118038/4853399a-e52d-42f3-99f4-535b78da1ebf)

**Comment by CVHub520** [comment]
Hello @kenwaytis,

Could you please share your annotated results with me? You can do so by pressing Ctrl+G and then sending me a screenshot.

**Comment by kenwaytis** [comment]
Hello @CVHub520,

Of course, I can share my results. However, I have already found a solution through experimentation.

**Solution**:
1. When you want to export annotation files in YOLO-Pose format, you need to set the **group_id** of the **detection box** and **keypoints** for the same object to the same number, as shown in the image below:
![2024-06-24_10-59](https://github.com/CVHub520/X-AnyLabeling/assets/40118038/4917dc4d-2774-4377-b3da-b150ded570fc)
2. The annotation points on the right will then look like this:
![2024-06-24_10-58](https://github.com/CVHub520/X-AnyLabeling/assets/40118038/1642574d-1229-452a-8b18-e387e7e570b8)
3. In the automatically saved standard file (json), the `group_id` is not empty.

---

In summary, the YOLO-Pose format needs to know which detection box the keypoints belong to, so you need to specify the `group_id` during annotation.

It might be because of my limited experience that I missed this point. I suggest that you consider adding the following guidance in the documentation:
>  "To export in YOLO-Pose format, you need to specify the `group_id` for each group (detection box and its keypoints) during annotation, so that the relationship can be understood during export."

If possible, I can submit a PR to add detailed instructions in `doc/language/user_guide.md`. Thank you very much for your time.

---

**Comment by CVHub520** [comment]
Hey! Thank you for reaching out and for sharing your findings! It's great to hear that you've resolved the issue with exporting annotations in YOLO-Pose format.

Your solution is clear and concise, and it will undoubtedly be helpful for others who might encounter a similar problem. That's right, the `group_id` is indeed a critical element in ensuring that the keypoints are correctly associated with their respective detection boxes.

Your suggestion to update the documentation with this information is very much appreciated. It will certainly improve the user experience by providing more detailed guidance on the annotation process for YOLO-Pose format exports.

I'd be delighted to accept a PR from you with the added instructions in `doc/language/user_guide.md`. Please feel free to submit it, and I'll review it as soon as possible. Your contribution to the project is invaluable, and I thank you in advance for helping to make our documentation even better.

Looking forward to your PR!

Best.

**Comment by kenwaytis** [comment]
> Hey! Thank you for reaching out and for sharing your findings! It's great to hear that you've resolved the issue with exporting annotations in YOLO-Pose format.嘿！感谢您伸出援手并分享您的发现！很高兴听到您已经解决了以 YOLO-Pose 格式导出注释的问题。
> 
> Your solution is clear and concise, and it will undoubtedly be helpful for others who might encounter a similar problem. That's right, the `group_id` is indeed a critical element in ensuring that the keypoints are correctly associated with their respective detection boxes.您的解决方案简洁明了，无疑会对可能遇到类似问题的其他人有所帮助。没错，这 `group_id` 确实是确保关键点与其各自的检测框正确关联的关键因素。
> 
> Your suggestion to update the documentation with this information is very much appreciated. It will certainly improve the user experience by providing more detailed guidance on the annotation process for YOLO-Pose format exports.非常感谢您建议使用此信息更新文档。它肯定会通过提供有关 YOLO-Pose 格式导出的注释过程的更详细指导来改善用户体验。
> 
> I'd be delighted to accept a PR from you with the added instructions in `doc/language/user_guide.md`. Please feel free to submit it, and I'll review it as soon as possible. Your contribution to the project is invaluable, and I thank you in advance for helping to make our documentation even better.我很高兴接受您的 PR 并在 `doc/language/user_guide.md` .请随时提交，我会尽快审核。您对项目的贡献是无价的，我提前感谢您帮助我们的文档变得更好。
> 
> Looking forward to your PR!期待您的公关！
> 
> Best. 最好。



> Hey! Thank you for reaching out and for sharing your findings! It's great to hear that you've resolved the issue with exporting annotations in YOLO-Pose format.
> 
> Your solution is clear and concise, and it will undoubtedly be helpful for others who might encounter a similar problem. That's right, the `group_id` is indeed a critical element in ensuring that the keypoints are correctly associated with their respective detection boxes.
> 
> Your suggestion to update the documentation with this information is very much appreciated. It will certainly improve the user experience by providing more detailed guidance on the annotation process for YOLO-Pose format exports.
> 
> I'd be delighted to accept a PR from you with the added instructions in `doc/language/user_guide.md`. Please feel free to submit it, and I'll review it as soon as possible. Your contribution to the project is invaluable, and I thank you in advance for helping to make our documentation even better.
> 
> Looking forward to your PR!
> 
> Best.

#472 
Thank you very much for your attitude towards improving the documentation. Best regards.

**Comment by ddongcui** [comment]
I find a potential reason for this error. When I load the pre-annotated object detection Txt  files and just annotate part of images with keypoints format， then I want to export to pose annotations， the program throws the same error.

**Comment by kenwaytis** [comment]
> I find a potential reason for this error. When I load the pre-annotated object detection Txt files and just annotate part of images with keypoints format， then I want to export to pose annotations， the program throws the same error.

@ddongcui 

Hello,

Regarding the point you mentioned about "annotating only parts of the image with key points in a key points format," I can provide my perspective.

When YOLO-POSE was designed, its task objective was to first perform bounding box detection and then detect key points within the bounding box. Therefore, there is no direct annotation of key points on the original image or direct detection of key points by the model on the image.

This means that each key point or group of key points is associated with a bounding box. This is why we need a `group_id` during annotation to indicate the relationship between the key points and the bounding box.

If this is still unclear, you can refer to the official annotation and prediction output formats for YOLO-POSE from Ultralytics. You will see that each line of data follows the format:

> bounding box x, bounding box y, bounding box width, bounding box height, bounding box confidence, key point 1 x, key point 1 y, key point 1 visibility (or confidence), key point 2 x, key point 2 y, key point 2 visibility, ... 
(visibility can sometimes be omitted).

**Comment by hypochlorous** [comment]
if you have finished mark,There is a script to update the group id;


Idw how to modify the completed keypoint annotation information in the software(windows),i try to modified one point,it wont apply to other file seems a bit troublesome.

```python
import json
import os

# Group ID mapping
group_id_mapping = {
    "LU_box": 1, "LM_box": 2, "LL_box": 3, "RL_box": 4, "RM_box": 5, "RU_box": 6,
    "LU_point1": 1, "LU_point2": 1, "LU_point3": 1,
    "LM_point1": 2, 
    "LL_point1": 3, "LL_point2": 3, "LL_point3": 3,
    "RL_point1": 4, "RL_point2": 4, "RL_point3": 4,
    "RM_point1": 5, 
    "RU_point0": 6, "RU_point1": 6, "RU_point2": 6, "RU_point3": 6
}

input_folder = "pathto/jsonfolder"

for filename in os.listdir(input_folder):

    if filename.endswith('.json'):
        json_path = os.path.join(input_folder, filename)
        
        # read every single json file. update group_id by mapping
        with open(json_path, 'r', encoding='utf-8') as file:
            json_data = json.load(file)

        for shape in json_data['shapes']:
            label = shape['label']
            if label in group_id_mapping:
                shape['group_id'] = group_id_mapping[label]

        with open(json_path, 'w', encoding='utf-8') as file:
            json.dump(json_data, file, indent=4)

print("JSON File update")
```

**Comment by hypochlorous** [comment]
> > I find a potential reason for this error. When I load the pre-annotated object detection Txt files and just annotate part of images with keypoints format， then I want to export to pose annotations， the program throws the same error.
> 
> @ddongcui
> 
> Hello,
> 
> Regarding the point you mentioned about "annotate part of images with keypoints format," I can provide my perspective.
> 
> When YOLO-POSE was designed, its task objective was to first perform bounding box detection and then detect key points within the bounding box. Therefore, there is no direct annotation of key points on the original image or direct detection of key points by the model on the image.
> 
> This means that each key point or group of key points is associated with a bounding box. This is why we need a `group_id` during annotation to indicate the relationship between the key points and the bounding box.
> 
> If this is still unclear, you can refer to the official annotation and prediction output formats for YOLO-POSE from Ultralytics. You will see that each line of data follows the format:
> 
> > bounding box x, bounding box y, bounding box width, bounding box height, bounding box confidence, key point 1 x, key point 1 y, key point 1 visibility (or confidence), key point 2 x, key point 2 y, key point 2 visibility, ...
> > (visibility can sometimes be omitted).

@kenwaytis  hello,

some people said: " Do not use yolopose on multi class keypoint detection. If there are multiple types objects, you'd better train multiple models "

Is this statement correct?On the official Ultralytics website, we can only find examples of a single type of task.

My task is to find the key points of several parts of a huge object, so I first perform object detection on the key parts and locate the key points within this patch. Should it works?
